File: 01 CLASE ENTERO/js/cadenas.py
class Cadena:
    def __init__(self, cadena):
        self.cadena = cadena

    # Obtener la cadena
    def get_cadena(self):
        return self.cadena

    # Concatenar otra cadena
    def concatenar(self, otra_cadena):
        self.cadena += otra_cadena

    # Invertir la cadena
    def invertir(self):
        self.cadena = self.cadena[::-1]

    # Verificar si es vacía
    def es_vacia(self):
        return len(self.cadena) == 0

    # Verificar si contiene una subcadena
    def contiene(self, subcadena):
        return subcadena in self.cadena

    # Verificar si comienza con un prefijo
    def comienza_con(self, prefijo):
        return self.cadena.startswith(prefijo)

    # Verificar si termina con un sufijo
    def termina_con(self, sufijo):
        return self.cadena.endswith(sufijo)

    # Obtener la longitud de la cadena
    def longitud(self):
        return len(self.cadena)

    # Convertir a mayúsculas
    def a_mayusculas(self):
        return self.cadena.upper()

    # Convertir a minúsculas
    def a_minusculas(self):
        return self.cadena.lower()

# Crear una instancia de la clase Cadena
cadena = Cadena('')

# Función para ejecutar el método correspondiente
def ejecutar(metodo, valor_input=''):
    resultado = ''

    if metodo == 'get_cadena':
        resultado = f"Cadena: {cadena.get_cadena()}"
    elif metodo == 'concatenar':
        cadena.concatenar(valor_input)
        resultado = f"Cadena actual: {cadena.get_cadena()}"
    elif metodo == 'invertir':
        cadena.invertir()
        resultado = f"Cadena invertida: {cadena.get_cadena()}"
    elif metodo == 'es_vacia':
        resultado = f"¿Es vacía?: {cadena.es_vacia()}"
    elif metodo == 'contiene':
        resultado = f"¿Contiene '{valor_input}'?: {cadena.contiene(valor_input)}"
    elif metodo == 'comienza_con':
        resultado = f"¿Comienza con '{valor_input}'?: {cadena.comienza_con(valor_input)}"
    elif metodo == 'termina_con':
        resultado = f"¿Termina con '{valor_input}'?: {cadena.termina_con(valor_input)}"
    elif metodo == 'longitud':
        resultado = f"Longitud: {cadena.longitud()}"
    elif metodo == 'a_mayusculas':
        resultado = f"A mayúsculas: {cadena.a_mayusculas()}"
    elif metodo == 'a_minusculas':
        resultado = f"A minúsculas: {cadena.a_minusculas()}"
    else:
        resultado = 'Método no reconocido.'

    return resultado

# Ejemplo de uso
print(ejecutar('concatenar', 'Hola'))
print(ejecutar('concatenar', ' Mundo'))
print(ejecutar('invertir'))
print(ejecutar('es_vacia'))
print(ejecutar('contiene', 'Hola'))
print(ejecutar('comienza_con', 'H'))
print(ejecutar('termina_con', 'o'))
print(ejecutar('longitud'))
print(ejecutar('a_mayusculas'))
print(ejecutar('a_minusculas'))

File: 02 CLASE CADENAS/js/cadenas.js
class Cadena {
    constructor(cadena) {
      this.cadena = cadena;
    }
  
    getCadena() { //Obtiene el valor asignado a la cadena
      return this.cadena;
    }
  
    // Manipulación de cadenas
    concatenar(otraCadena) {  //Unir una cadena con otra
      this.cadena += otraCadena;
    }
  
    invertir() { //Invertir los caracteres de la cadena
      let cadenaInvertida = '';
      for (let i = this.cadena.length - 1; i >= 0; i--) {
        cadenaInvertida += this.cadena[i];
      }
      this.cadena = cadenaInvertida;
    }
  
    // Comparaciones
    esVacia() { //Compara si la cadena no tiene caracteres
      return this.cadena.length === 0;
    }
  
    contiene(subcadena) { //Compara si la subCadena existe dentro de la cadena
      return this.cadena.includes(subcadena);
    }
  
    comienzaCon(prefijo) { //Compara si la cadena comienza con el prefijo "caracter"
      return this.cadena.startsWith(prefijo);
    }
  
    terminaCon(sufijo) { //Compara si la cadena termina con el sufijo "caracter"
      return this.cadena.endsWith(sufijo);
    }
  
    // Otras operaciones
    longitud() { //Obtiene la cantidad de caracteres que tiene la cadena
      return this.cadena.length;
    }
  
    aMayusculas() { //Obtiene la cadena transformada en MAYUSCULA
      return this.cadena.toUpperCase();
    }
  
    aMinusculas() { //Obtiene la cadena transformada en minuscula
      return this.cadena.toLowerCase();
    }

    // 1. Transformación de números a números romanos
    static aNumeroRomano(numero) {
      const valores = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1];
      const simbolos = ["M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"];
      let resultado = '';

      for (let i = 0; i < valores.length; i++) {
          while (numero >= valores[i]) {
              resultado += simbolos[i];
              numero -= valores[i];
          }
      }
      return resultado;
    }

    // 2. Generación de texto aleatorio
    static generarTextoAleatorio(palabras) {
      const caracteres = 'abcdefghijklmnopqrstuvwxyz0123456789';
      let resultado = '';
      for (let i = 0; i < palabras; i++) {
          let palabra = '';
          const longitud = Math.floor(Math.random() * 5) + 3; // Palabras de 3 a 7 caracteres
          for (let j = 0; j < longitud; j++) {
              palabra += caracteres.charAt(Math.floor(Math.random() * caracteres.length));
          }
          resultado += palabra + ' ';
      }
      return resultado.trim();
    }

    // 3. Transformación de ASCII a Texto
    static asciiATexto(num) {
        if (num < 0 || num > 127) {
          throw new Error("El número debe estar entre 0 y 127.");
        }
      return String.fromCharCode(num); // Convierte el número a su representación ASCII
    }
  }
  
  let cadena = new Cadena('');
  
  function ejecutar(metodo) {
    const input = document.getElementById('inputCadena');
    const resultadoDiv = document.getElementById('resultado');
    const valorInput = input.value;
  
    switch (metodo) {
      case 'getCadena':
        resultadoDiv.innerHTML = `Cadena: ${cadena.getCadena()}`;
        break;
      case 'concatenar':
        cadena.concatenar(valorInput);
        resultadoDiv.innerHTML = `Cadena actual: ${cadena.getCadena()}`;
        break;
      case 'invertir':
        cadena.invertir();
        resultadoDiv.innerHTML = `Cadena invertida: ${cadena.getCadena()}`;
        break;
      case 'esVacia':
        resultadoDiv.innerHTML = `¿Es vacía?: ${cadena.esVacia()}`;
        break;
      case 'contiene':
        resultadoDiv.innerHTML = `¿Contiene '${valorInput}'?: ${cadena.contiene(valorInput)}`;
        break;
      case 'comienzaCon':
        resultadoDiv.innerHTML = `¿Comienza con '${valorInput}'?: ${cadena.comienzaCon(valorInput)}`;
        break;
      case 'terminaCon':
        resultadoDiv.innerHTML = `¿Termina con '${valorInput}'?: ${cadena.terminaCon(valorInput)}`;
        break;
      case 'longitud':
        resultadoDiv.innerHTML = `Longitud: ${cadena.longitud()}`;
        break;
      case 'aMayusculas':
        resultadoDiv.innerHTML = `A mayúsculas: ${cadena.aMayusculas()}`;
        break;
      case 'aMinusculas':
        resultadoDiv.innerHTML = `A minúsculas: ${cadena.aMinusculas()}`;
        break;
        case 'aNumeroRomano':
          resultadoDiv.innerHTML = `Número Romano: ${Cadena.aNumeroRomano(parseInt(valorInput))}`;
          break;
      case 'generarTextoAleatorio':
          resultadoDiv.innerHTML = `Texto Aleatorio: ${Cadena.generarTextoAleatorio(parseInt(valorInput))}`;
          break;
      case 'asciiATexto':
          resultadoDiv.innerHTML = `Texto ASCII: ${Cadena.asciiATexto(parseInt(valorInput))}`;
          break;
      default:
        resultadoDiv.innerHTML = 'Método no reconocido.';
    }
}